tests/interfaces/test_batch_autograd_qnode.py

```diff
@@ -0,0 +1,1238 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Integration tests for using the autograd interface with a QNode"""
+import pytest
+from pennylane import numpy as np
+
+import pennylane as qml
+from pennylane.beta import qnode, QNode
+from pennylane.tape import JacobianTape
+
+qubit_device_and_diff_method = [
+    ["default.qubit", "finite-diff", "backward"],
+    ["default.qubit", "parameter-shift", "backward"],
+    ["default.qubit", "backprop", "forward"],
+    ["default.qubit", "adjoint", "forward"],
+    ["default.qubit", "adjoint", "backward"],
+]
+
+
+@pytest.mark.parametrize("dev_name,diff_method,mode", qubit_device_and_diff_method)
+class TestQNode:
+    """Test that using the QNode with Autograd integrates with the PennyLane stack"""
+
+    def test_nondiff_param_unwrapping(self, dev_name, diff_method, mode, mocker):
+        """Test that non-differentiable parameters are correctly unwrapped
+        to NumPy ndarrays or floats (if 0-dimensional)"""
+        if diff_method != "parameter-shift":
+            pytest.skip("Test only supports parameter-shift")
+
+        dev = qml.device("default.qubit", wires=1)
+
+        @qnode(dev, interface="autograd", diff_method="parameter-shift")
+        def circuit(x, y):
+            qml.RX(x[0], wires=0)
+            qml.Rot(*x[1:], wires=0)
+            qml.RY(y[0], wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        x = np.array([0.1, 0.2, 0.3, 0.4], requires_grad=False)
+        y = np.array([0.5], requires_grad=True)
+
+        param_data = []
+
+        def mock_apply(*args, **kwargs):
+            for op in args[0]:
+                param_data.extend(op.data.copy())
+
+        mocker.patch.object(dev, "apply", side_effect=mock_apply)
+        circuit(x, y)
+        assert param_data == [0.1, 0.2, 0.3, 0.4, 0.5]
+        assert not any(isinstance(p, np.tensor) for p in param_data)
+
+        # test the jacobian works correctly
+        param_data = []
+        qml.grad(circuit)(x, y)
+        assert param_data == [
+            0.1,
+            0.2,
+            0.3,
+            0.4,
+            0.5,
+            0.1,
+            0.2,
+            0.3,
+            0.4,
+            0.5 + np.pi / 2,
+            0.1,
+            0.2,
+            0.3,
+            0.4,
+            0.5 - np.pi / 2,

```

Comment:
Just to make sure I understand what's happened here, instead of executing anything, mock apply just appends to the parameter list, *including* when we call `grad`? So here what's been appended are the two sets of parameters for parameter shift, where only the `y` parameter is shifted since `x` has `requires_grad=False`?

Reply:
yep exactly!! Blame past Josh for this, I wrote this test last year 😬 

I actually probably wouldn't have written it like this year, I feel like my mocking ability has declined lol